### PR TITLE
SHARD-2216 : everything from shard 2234, shard 2235 and shard 2236 in 1 pr

### PR DESCRIPTION
**Pull request by devendra-shardeum**
### **PR Type**
Enhancement, Tests

___

### **Description**
- Refactored account hash verification to be fully async and robust.

- Improved error handling and logging in account hash and receipt verification.

- Added new DB query for counting InitNetwork receipts between cycles.

- Updated and expanded unit tests for async and error scenarios.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>calculateAccountHash.ts</strong><dd><code>Refactor account hash calculation and verification to async</code></dd></summary>
<hr>

src/shardeum/calculateAccountHash.ts

<li>Refactored account hash and verification functions to async.<br> <li> Improved error handling and logging for account hash calculation.<br> <li> Added detailed JSDoc comments for exported functions.<br> <li> Split and clarified global vs non-global account verification logic.

</details>

  </td>
  <td><a href="https://github.com/shardeum/archiver/pull/233/files#diff-3917e7bd2a59e99c0a792c8725d95eb5bb43504675b3a48511617f02b20ec862">+94/-53</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>verifyGlobalTxReceipt.ts</strong><dd><code>Async global receipt verification and InitNetwork count check</code></dd></summary>
<hr>

src/shardeum/verifyGlobalTxReceipt.ts

<li>Refactored <code>verifyGlobalTxAccountChange</code> to async.<br> <li> Integrated new DB query for InitNetwork receipt count.<br> <li> Improved hash comparison logic and error handling.

</details>

  </td>
  <td><a href="https://github.com/shardeum/archiver/pull/233/files#diff-9aed88fb73efd157c26b57ec9af61cad2ec0c9c0fde80bff01b82d1002645e89">+8/-8</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>Collector.ts</strong><dd><code>Async account hash verification and state storage fix</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/Data/Collector.ts

<li>Made account hash verification async in receipt verification.<br> <li> Fixed account object property access in account state storage.

</details>

  </td>
  <td><a href="https://github.com/shardeum/archiver/pull/233/files#diff-186c711dbe48d8ffd4c6488e2d29090dae64abd946a5cf185a549af2c556a72e">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>receipts.ts</strong><dd><code>Add InitNetwork receipt count query with error handling</code>&nbsp; &nbsp; </dd></summary>
<hr>

src/dbstore/receipts.ts

<li>Added async function to count InitNetwork receipts between cycles.<br> <li> Improved error handling and logging for new query.

</details>

  </td>
  <td><a href="https://github.com/shardeum/archiver/pull/233/files#diff-e644b6b6f429a07840b014023c9b7412f31a9821c05283431844d275fa4b9a81">+40/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>calculateAccountHash.test.ts</strong><dd><code>Update and expand async account hash verification tests</code>&nbsp; &nbsp; </dd></summary>
<hr>

test/unit/src/calculateAccountHash.test.ts

<li>Updated tests for async account hash verification.<br> <li> Added/modified tests for error handling and async scenarios.<br> <li> Improved test coverage for edge cases and exceptions.

</details>

  </td>
  <td><a href="https://github.com/shardeum/archiver/pull/233/files#diff-e704b29220b010ebca54803ccf4c15cabbe9c6da4724ccc5401cb6cddfade72e">+22/-35</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>verifyGlobalTxReceipt.test.ts</strong><dd><code>Update tests for async global receipt verification</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

test/unit/src/shardeum/verifyGlobalTxReceipt.test.ts

<li>Updated tests for async global receipt verification.<br> <li> Adjusted tests for new async and error handling logic.

</details>

  </td>
  <td><a href="https://github.com/shardeum/archiver/pull/233/files#diff-e96e4dccda2b45719c3c973696a83907a97a7115bce53cc54f7505eb1251ec0b">+4/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

**Review by aniketdivekar — DISMISSED** [review]

**Review by urnotsam — APPROVED** [review]

File: src/Data/Collector.ts [hunk 1]
```diff
@@ -849,39 +849,111 @@ export const storeReceiptData = async (
...
+            const account = receipt.afterStates.find((acc) => acc.accountId === accountId);
+            if (!account) {
+              Logger.mainLogger.error('Account not found in afterStates', txId, accountId);
+              continue;
+            }
+      
+            const accObj: Account.AccountsCopy = {
+              accountId: account.accountId,
+              data: account.data,
+              timestamp: account.data.timestamp,
```
Comment (afostr) [inline src/Data/Collector.ts:924]: I think this will mean we later compare account.data.timestamp to account.data['timestamp']

File: src/Data/Collector.ts [hunk 2]
```diff
@@ -849,39 +849,111 @@ export const storeReceiptData = async (
...
+          globalReceiptValidationErrors = verifyPayload(AJVSchemaEnum.GlobalTxReceipt, receipt?.signedReceipt)
+          
+          if (!globalReceiptValidationErrors) {
+
+            for (const account of afterStates) {
+
+              const accObj: Account.AccountsCopy = {
+                accountId: config.globalNetworkAccount, // for global tx type receipts fixing accountID to 1000000000000000000000000000000000000000000000000000000000000001 for now
+                data: account.data,
+                timestamp: account.data.timestamp,
```
Comment (afostr) [inline src/Data/Collector.ts:867]: this means we will compare account.data.timestamp to account.data['timestamp'] below

File: src/shardeum/verifyGlobalTxReceipt.ts [hunk 3]
```diff
@@ -49,18 +49,21 @@ export enum InternalTXType {
...
-): boolean => {
+): Promise<boolean> => {
   try {
     const signedReceipt = receipt.signedReceipt as P2P.GlobalAccountsTypes.GlobalTxReceipt
     const internalTx = signedReceipt.tx.value as SetGlobalTxValue
 
     if (internalTx.internalTXType === InternalTXType.InitNetwork) {
       // Refer to https://github.com/shardeum/shardeum/blob/89db23e1d4ffb86b4353b8f37fb360ea3cd93c5b/src/index.ts#L2334
       // no need to do anything, as it is network account creation
+      let count = await queryInitNetworkReceiptCountBetweenCycles(1, 5)
```
Comment (afostr) [inline src/shardeum/verifyGlobalTxReceipt.ts:64]: what happens when cycle > 5.
seems like we should check that first for quick reject.
I don't think we need to allow for the case where we are past cycle 5 and don't have a network account yet.

File: src/shardeum/calculateAccountHash.ts [hunk 4]
```diff
@@ -22,68 +22,58 @@ export enum AccountType {
...
+  try {
+    // Remove the existing hash property from the account object
+    delete account.hash
+
+    // Calculate a new hash based on the account's data and assign it to the hash property
     account.hash = crypto.hashObj(account)
     return account.hash
+  } catch (error) {
+    console.error('Error calculating account-specific hash:', error)
+    throw new Error('Failed to calculate account-specific hash')
```
Comment (afostr) [inline src/shardeum/calculateAccountHash.ts:48]: This is a similar situation to my comments on throwing an error in the equivalent code in shardeum.   That said, the surface area here should be much smaller, so I am fine as long is enough code review is put into checking.

**Review by afostr — CHANGES_REQUESTED** [review]
changes requested

File: src/Data/Collector.ts (hunk 2 above)
Comment (devendra-shardeum) [reply src/Data/Collector.ts:867]: yes.. since account.data is now involved in hash calculation we can trust it.. and fortunately it's present inside the account.data and can be compared and used.. using account.timestamp isn't a good idea imo since it's not trustable

File: src/Data/Collector.ts (hunk 1 above)
Comment (devendra-shardeum) [reply src/Data/Collector.ts:924]: yes it's old logic anyways

File: src/shardeum/calculateAccountHash.ts (hunk 4 above)
Comment (devendra-shardeum) [reply src/shardeum/calculateAccountHash.ts:48]: archivers won't crash because of global error handlers but can reach an unknown state during it's running if it encounter errors.. but yeah surface area is small and that receipt will be ignored and it'll continue to the next receipt

File: src/shardeum/verifyGlobalTxReceipt.ts (hunk 3 above)
Comment (devendra-shardeum) [reply src/shardeum/verifyGlobalTxReceipt.ts:64]: initNetwork will be sent only once and not later during network running and i think  `internalTx.internalTXType === InternalTXType.InitNetwork` will work as a quick check imo and db query won't be running every time.. and in case when malicious validator send these type of tx receipt then in those cases as well cycle can't be used to have a quick check since receipt.cycle can't be trusted and it's not present in other parts of receipt for initNetwork where i can extract it and use it.

File: src/shardeum/calculateAccountHash.ts (hunk 4 above)
Comment (devendra-shardeum) [reply src/shardeum/calculateAccountHash.ts:48]: also few comments/concern are same as mentioned in other pr as well https://github.com/shardeum/shardeum/pull/459#discussion_r2048577562

**Review by afostr — APPROVED** [review]
looks good
